### PR TITLE
modify the modification req handler

### PR DESCRIPTION
**Pull request by ianchen0119**
[UPF PR #47 ](https://github.com/free5gc/upf/pull/47)

According to TS 29.244:
- 6.3.3 PFCP Session Modification Procedure

- 7.2.2.4.2 Conditions for Sending SEID=0 in PFCP Header

Even if the session hasn't been found when processing the N4 modification request, UPF should send the N4 modification 
response with cause `PFCP_CAUSE_SESSION_CONTEXT_NOT_FOUND` and SEID = 0 to SMF still.

### CODE REVIEW

File: src/n4/n4_pfcp_build.c [hunk 1]
```diff
@@ -69,21 +69,32 @@ Status UpfN4BuildSessionModificationResponse(Bufblk **bufBlkPtr, uint8_t type,
...
     response = &pfcpMessage.pFCPSessionModificationResponse;
     memset(&pfcpMessage, 0, sizeof(pfcpMessage));
 
     /* cause */
     response->cause.presence = 1;
-    cause = PFCP_CAUSE_REQUEST_ACCEPTED;
-    response->cause.value = &cause;
-    response->cause.len = 1;
-
+    if (!sessionIsExist) {
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_build.c:78]:
```
if (sessionExisted) {
    cause = PFCP_CAUSE_REQUEST_ACCEPTED;
} else {
    cause = PFCP_CAUSE_SESSION_CONTEXT_NOT_FOUND;
}
response->cause.value = &cause;
response->cause.len = sizeof(cause);
```

File: src/n4/n4_pfcp_build.c [hunk 2]
```diff
@@ -69,21 +69,32 @@ Status UpfN4BuildSessionModificationResponse(Bufblk **bufBlkPtr, uint8_t type,
     PfcpMessage pfcpMessage;
     PFCPSessionModificationResponse *response = NULL;
     uint8_t cause;
-
+    bool sessionIsExist = (session)?(true):(false);
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_build.c:72]: rename to `sessionExisted`

File: src/n4/n4_pfcp_build.c [hunk 3]
```diff
@@ -69,21 +69,32 @@ Status UpfN4BuildSessionModificationResponse(Bufblk **bufBlkPtr, uint8_t type,
...
+        response->cause.value = &cause;
+        response->cause.len = 1;
+    }
+    
     /* TODO: Set Offending IE, Create PDR, Load Control Information, Overload Control Information, Usage Report, Failed Rule ID, Additional Usage Reports Information, Created/Updated Traffic Endpoint */
 
     pfcpMessage.header.type = type;
     pfcpMessage.header.seidP = 1;
-    pfcpMessage.header.seid = session->smfSeid;
+    if (!sessionIsExist) {
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_build.c:92]:
```
    if (sessionExisted) {
        pfcpMessage.header.seid = session->smfSeid;
    } else {
        /* TS 29.244 Clause 7.2.2.4.2 Conditions for Sending SEID=0 in PFCP Header */
        pfcpMessage.header.seid = 0;
    }
```

File: src/n4/n4_pfcp_handler.c [hunk 4]
```diff
@@ -1188,109 +1188,111 @@ Status UpfN4HandleSessionEstablishmentRequest(UpfSession *session, PfcpXact *pfc
...
     PfcpHeader header;
     Bufblk *bufBlk;
 
-    /* Create FAR */
-    for (int i = 0; i < sizeof(request->createFAR) / sizeof(CreateFAR); i++) {
-        if (request->createFAR[i].presence) {
-            status = UpfN4HandleCreateFar(session, &request->createFAR[i]);
-            UTLT_Assert(status == STATUS_OK, return STATUS_ERROR,
-                        "Modification: Create FAR error");
+    if (session) {
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_handler.c:1198]:
```
if (!session) {
    UTLT_Warning("Session is not found")
} else {
    ...
}
```

File: src/n4/n4_pfcp_handler.c [hunk 5]
```diff
@@ -1188,109 +1188,111 @@ Status UpfN4HandleSessionEstablishmentRequest(UpfSession *session, PfcpXact *pfc
 
 Status UpfN4HandleSessionModificationRequest(UpfSession *session, PfcpXact *xact,
                                              PFCPSessionModificationRequest *request) {
-    UTLT_Assert(session, return STATUS_ERROR, "Session error");
+    UTLT_Assert(session, NOOP, "Session is not found");
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_handler.c:1191]: remove this

**Review by free5gc-org — COMMENTED** [review]
Should do the same check in `UpfN4HandleSessionDeletionRequest`

**Comment by ianchen0119** [comment]
> Should do the same check in `UpfN4HandleSessionDeletionRequest`

Done!

Please help to check when you feel free,
Thanks!

File: src/n4/n4_pfcp_handler.c [hunk 6]
```diff
@@ -1336,7 +1343,7 @@ Status UpfN4HandleSessionDeletionRequest(UpfSession *session, PfcpXact *xact,
     header.seid = session->smfSeid;
 
     status = UpfN4BuildSessionDeletionResponse(&bufBlk, header.type,
-                                               session, request);
+                                               session, request, sessionExisted);
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_handler.c:1346]: remove `sessionExisted`

File: src/n4/n4_pfcp_build.c [hunk 7]
```diff
@@ -69,21 +69,30 @@ Status UpfN4BuildSessionModificationResponse(Bufblk **bufBlkPtr, uint8_t type,
...
-    cause = PFCP_CAUSE_REQUEST_ACCEPTED;
     response->cause.value = &cause;
-    response->cause.len = 1;
-
+    response->cause.len = sizeof(cause);
+    
     /* TODO: Set Offending IE, Create PDR, Load Control Information, Overload Control Information, Usage Report, Failed Rule ID, Additional Usage Reports Information, Created/Updated Traffic Endpoint */
 
     pfcpMessage.header.type = type;
     pfcpMessage.header.seidP = 1;
```
Comment (free5gc-org) [inline src/n4/n4_pfcp_build.c:89]: remove this line. only type need to be assigned.

File: src/n4/n4_pfcp_handler.c [hunk 8]
```diff
@@ -1188,109 +1188,112 @@ Status UpfN4HandleSessionEstablishmentRequest(UpfSession *session, PfcpXact *pfc
...
+            if (request->removePDR[i].presence) {
+                UTLT_Assert(request->removePDR[i].pDRID.presence == 1, ,
+                            "[PFCP] PdrId in removePDR not presence!");
+                status = UpfN4HandleRemovePdr(session, *(uint16_t*)
+                                            request->removePDR[i].pDRID.value);
+                UTLT_Assert(status == STATUS_OK, return STATUS_ERROR,
+                            "Modification: Remove PDR error");
+            }
         }
     }
```
[the remark this answers is not in the document]
Comment (free5gc-org) [reply src/n4/n4_pfcp_handler.c:1299]: Build hdr in PfcpXactUpdateTx()

File: src/n4/n4_pfcp_handler.c [hunk 9]
```diff
@@ -1336,7 +1343,7 @@ Status UpfN4HandleSessionDeletionRequest(UpfSession *session, PfcpXact *xact,
     header.seid = session->smfSeid;
```
[the remark this answers is not in the document]
Comment (free5gc-org) [reply src/n4/n4_pfcp_handler.c:1343]: Build hdr in PfcpXactUpdateTx()

File: src/n4/n4_dispatcher.c [hunk 10]
```diff
@@ -75,7 +75,7 @@ void UpfDispatcher(const Event *event) {
                 session = UpfSessionFindBySeid(pfcpMessage->header.seid);
             }
 
-            UTLT_Level_Assert(LOG_WARNING, session, goto freeBuf,
+            UTLT_Level_Assert(LOG_WARNING, session, NOOP,
```
Comment (free5gc-org) [inline src/n4/n4_dispatcher.c:78]:
```
UTLT_Level_Assert(LOG_WARNING,
    session || pfcpMessage->header.type != PFCP_SESSION_ESTABLISHMENT_REQUEST, goto freeBuf,
    "can't establish session");
```

File: src/n4/n4_dispatcher.c [hunk 11]
```diff
@@ -75,7 +75,7 @@ void UpfDispatcher(const Event *event) {
...
                         "do not find / establish session");
 
             if (pfcpMessage->header.type != PFCP_SESSION_REPORT_RESPONSE) {
```
Comment (free5gc-org) [inline src/n4/n4_dispatcher.c:81]:
```
            if (pfcpMessage->header.type != PFCP_SESSION_REPORT_RESPONSE && session) {
                session->pfcpNode = upf;
            }
            if (session) {
                status = PfcpXactReceive(session->pfcpNode,
                                         &pfcpMessage->header, &xact);
                UTLT_Assert(status == STATUS_OK, goto freeBuf, "");
            }
```

Comment (free5gc-org) [reply src/n4/n4_dispatcher.c:81]: xact will be NULL if session is NULL

**Review by free5gc-org — COMMENTED** [review]
Need to write another send Pfcp message function which will not rely on xact.

**Pull request closed without merging**